op-challenger/fault/game.go

```diff
@@ -1,13 +1,151 @@
 package fault
 
+import (
+	"errors"
+
+	"github.com/ethereum/go-ethereum/common"
+)
+
+var (
+	// ErrClaimExists is returned when a claim already exists in the game state.
+	ErrClaimExists = errors.New("claim exists in game state")
+
+	// ErrClaimNotFound is returned when a claim does not exist in the game state.
+	ErrClaimNotFound = errors.New("claim not found in game state")
+)
+
 // Game is an interface that represents the state of a dispute game.
 type Game interface {
-	// Put adds a claim into the game state and returns its parent claim.
-	Put(claim Claim) (Claim, error)
+	// Put adds a claim into the game state.
+	Put(claim Claim) error
 
 	// ClaimPairs returns a list of claim pairs.
 	ClaimPairs() []struct {
 		claim  Claim
 		parent Claim
 	}
 }
+
+// Node is a node in the game state tree.
+type Node struct {
+	self     Claim
+	children []Node
+}
+
+// NodeKey is the key for a given list of [Claim] children.
+type NodeKey struct {
+	// Position is embedded to pass through its methods.
+	Position
+	// Value is the claim hash.
+	Value common.Hash
+}
+
+// FromClaim creates a [NodeKey] from a given [Claim].
+func FromClaim(claim *Claim) NodeKey {
+	return NodeKey{
+		Position: claim.Position,
+		Value:    claim.Value,
+	}
+}
+
+// gameState is a struct that represents the state of a dispute game.
+// The game state implements the [Game] interface.
+type gameState struct {
+	nodes map[NodeKey]Node
```

Comment:
Why do we have a map of nodes? The root node is all we need in the game state to be able to find every other node.

Reply:
I think it allows for simple, easy access to nodes and their parents.

A root node _should_ allow us to find every other node **iff** the agent gracefully decides not confuse the state by first posting an intermediary node.

I think this is a straightforward approach, and don't see why we'd limit ourselves by only containing the root node when storage isn't the limiting factor here. Lookup speed is far more important and is O(1) with the mapping. With a rootnode it's Θ(d/2)

Reply:
> A root node should allow us to find every other node iff the agent gracefully decides not confuse the state by first posting an intermediary node.

This isn't something that happens - the contract requires that claims point to a parent claim (the root claim is initialized at contract creation time).

Reply:
Sure, it's still directly limiting though and provides zero benefits.